Add stress test for log forwarding

pvollan opened the pull request:
#### b076a8f96407ea38b1ad3126579bfba401d4097b
<pre>
Add stress test for log forwarding
<a href="https://bugs.webkit.org/show_bug.cgi?id=292538">https://bugs.webkit.org/show_bug.cgi?id=292538</a>
<a href="https://rdar.apple.com/problem/150678850">rdar://problem/150678850</a>

Reviewed by Sihui Liu.

Add stress test for log forwarding from the WebContent process. This test will emit many logs from 2 WebContent processes,
from both the main thread and a background thread simultaneously.

* Source/WebCore/platform/LogMessages.in:
* Source/WebCore/platform/Logging.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::emitWebCoreLogs const):
(WebCore::Internals::emitLogs const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Platform/LogClient.h:
* Source/WebKit/Scripts/generate-derived-log-sources.py:
(generate_message_receiver_implementations_file):
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::logOnBehalfOfWebContent):
* Source/WebKit/Shared/LogStream.h:
(WebKit::LogStream::logCount):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _forwardedLogsCount]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm: Added.
(TEST(WebKit, LogForwarding)):

Canonical link: <a href="https://commits.webkit.org/294619@main">https://commits.webkit.org/294619@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5620feb3c2a3fd300f0e74c5841cb5f67da9996f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102463 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22132 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12446 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/107624 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53100 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104502 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22440 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/30638 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77953 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34943 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105469 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17385 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92489 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58289 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/101937 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/17222 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10517 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52457 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87040 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10588 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109999 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/29595 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21824 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86935 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29959 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/88685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86527 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31356 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9077 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/23840 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16646 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/29523 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/34826 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29334 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/32657 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/30893 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -41,6 +41,8 @@
 
 namespace WebKit {
 
+unsigned LogStream::m_logCount = 0;
```
Maybe `static std::atomic<unsigned> globalLogCount { 0 };`

pvollan replied:
That is a good catch! I'll address these issues in the next patch.

Thanks for reviewing!

szewai commented on Source/WebKit/Shared/LogStream.h:
```diff
@@ -60,6 +62,7 @@ class LogStream : public IPC::StreamMessageReceiver {
     RefPtr<IPC::StreamServerConnection> m_logStreamConnection;
     int32_t m_pid { 0 };
     Markable<LogStreamIdentifier> m_logStreamIdentifier;
+    static unsigned m_logCount;
```
This seems unnecessary to be data member of `LogStream` if it's only used in `LogStream`; we may move definition of `logCount()` and this to `LogStream.cpp`?

pvollan replied:
The WKWebView testing SPI is calling `logCount()`, so I think we need to keep the `logCount()` method.

szewai replied:
We can keep `logCount()` declaration here but move its definition to LogStream.cpp, then we don't need `m_logCount ` here. i.e. `unsigned LogStream::logCount() { return globalLogCount; }` in LogStream.cpp

szewai commented on Source/WebKit/Shared/LogStream.cpp:
```diff
@@ -61,6 +63,8 @@ void LogStream::logOnBehalfOfWebContent(std::span<const uint8_t> logSubsystem, s
 {
     ASSERT(!isMainRunLoop());
 
+    m_logCount++;
```
If this is called on non-main runloop, should it be atomic?

szewai commented on Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm:
```diff
@@ -0,0 +1,98 @@
+    constexpr ASCIILiteral logMessageOnMainThread = "Log message on main thread"_s;
+
+    for (int i = 0; i < 32; i++) {
+        ASSERT_TRUE(emitWebCoreLogsFromMainThread(webView.get(), mainThreadLogCount));
```
We should use `EXPECT_TRUE` if this is non-fatal failure

szewai commented on Source/WebCore/testing/Internals.mm:
```diff
@@ -273,4 +275,32 @@ - (NSAttributedString *)_attributedStringForRange:(NSRange)range
+{
+    auto blockPtr = makeBlockPtr([logCount] {
+        for (unsigned i = 0; i < logCount; i++)
+            RELEASE_LOG_FORWARDABLE(Media, HTMLMEDIAELEMENT_CREATEMEDIAPLAYER, 0ull);
```
Can we define a forward log with `Testing` category and then add `LogStream::testingLogCount` to count only Testing category? This should make the API test more accurate (since we could wait count to reach 1024).

szewai commented on Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm:
```diff
@@ -0,0 +1,95 @@
+    EXPECT_TRUE(emitLogsFromMainThread(webView.get(), logMessageOnMainThread, mainThreadLogCount));
+    EXPECT_TRUE(emitLogsFromMainThread(webView2.get(), logMessageOnMainThread, mainThreadLogCount));
+
+    [webView loadRequest:[NSURLRequest requestWithURL:[NSBundle.test_resourcesBundle URLForResource:@"simple" withExtension:@"html"]]];
+    [webView2 loadRequest:[NSURLRequest requestWithURL:[NSBundle.test_resourcesBundle URLForResource:@"simple" withExtension:@"html"]]];
```
Why do we need to reload the views?

pvollan replied:
That is a good point! It is not required. I added it to test a realistic scenario, where we get logs related to loading at the same time.

szewai commented on Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm:
```diff
@@ -0,0 +1,95 @@
+    [webView2 loadRequest:[NSURLRequest requestWithURL:[NSBundle.test_resourcesBundle URLForResource:@"simple" withExtension:@"html"]]];
+
+    do {
+        if ([webView _forwardedLogsCountForTesting] >= 4 * (backgroundThreadLogCount + mainThreadLogCount))
```
Maybe:
```
unsigned expectedLogCount = (backgroundThreadLogCount + mainThreadLogCount) * 4;
while ([webView _forwardedLogsCountForTesting] < expectedLogCount)
    TestWebKitAPI::Util::spinRunLoop(1);
```

pvollan replied:
Thanks for reviewing!

pvollan committed:
Add stress test for log forwarding
https://bugs.webkit.org/show_bug.cgi?id=292538
rdar://problem/150678850

Reviewed by Sihui Liu.

Add stress test for log forwarding from the WebContent process. This test will emit many logs from 2 WebContent processes,
from both the main thread and a background thread simultaneously.

* Source/WebCore/platform/LogMessages.in:
* Source/WebCore/platform/Logging.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::emitWebCoreLogs const):
(WebCore::Internals::emitLogs const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Platform/LogClient.h:
* Source/WebKit/Scripts/generate-derived-log-sources.py:
(generate_message_receiver_implementations_file):
* Source/WebKit/Shared/LogStream.cpp:
(WebKit::LogStream::logOnBehalfOfWebContent):
* Source/WebKit/Shared/LogStream.h:
(WebKit::LogStream::logCount):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivateForTesting.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewTesting.mm:
(-[WKWebView _forwardedLogsCount]):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKit/LogForwarding.mm: Added.
(TEST(WebKit, LogForwarding)):

Canonical link: https://commits.webkit.org/294619@main
